### PR TITLE
correctness fixes for the autobatch blockstore

### PR DESCRIPTION
**Pull request by Stebalien**
1. Simplify shutdown and make it idempotent by using a context.
2. Make sure `Flush` actually _fully_ flushes if the previous flush failed.
3. Don't clear the flush batch if flushing fails.

### CODE REVIEW

File: blockstore/autobatch.go [hunk 1]
```diff
@@ -88,66 +86,85 @@ func (bs *AutobatchBlockstore) Put(ctx context.Context, blk block.Block) error {
...
-		bs.shutdownCh <- struct{}{}
+	// TODO: Prevent puts after we call this to avoid losing data.
+	bs.shutdown()
+	select {
+	case <-bs.doneCh:
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 
+	bs.doFlushLock.Lock()
```
Comment (arajasek) [inline blockstore/autobatch.go:165]: Is this just to make sure that any earlier claimers of the lock finish before we return?

**Review by arajasek — APPROVED** [review]

**Comment by Stebalien** [comment]
(forgot to create the channel)

File: blockstore/autobatch.go (hunk 1 above)
Comment (Stebalien) [reply blockstore/autobatch.go:166]: Concurrent Flush and Shutdown.

**Comment by jennijuju** [comment]
folllowing the PR template - a nice PR title should be created so our change log is pretty

**Pull request merged**